xbmc/platform/Platform.h

```diff
@@ -65,4 +95,65 @@ class CPlatform
    * Logs platform specific system info during application creation startup
    */
   virtual void PlatformSyslog() {}
+
+  /**\brief Get a platform service instance.
+   */
+  template<class T>
+  std::shared_ptr<T> GetService()
+  {
+    std::unique_lock<CCriticalSection> lock(m_critSection);
+    for (const auto& service : m_services)
+    {
+      if (service.get())
+      {
+        const auto& serviceRef = *service.get();
+        if (typeid(serviceRef) == typeid(T))
+          return std::static_pointer_cast<T>(service);
+      }
+    }
+
+    return nullptr;
+  }
+
+protected:
+  /**\brief Register a new platform service instance.
+   */
+  void RegisterService(const std::shared_ptr<IPlatformService>& service)
+  {
+    if (!service.get())
+      return;
+
+    const auto& serviceRef = *service.get();
+
+    std::unique_lock<CCriticalSection> lock(m_critSection);
+    for (const auto& s : m_services)
+    {
+      if (s.get())
+      {
+        const auto& sRef = *s.get();
+        if (typeid(sRef) == typeid(serviceRef))
+          return; // already registered
+      }
+    }
+
+    m_services.emplace_back(service);
+  }
+
+  /**\brief Deregister a platform service instance.
+   */
+  void DeregisterService(const std::type_info& typeInfo)
+  {
+    std::unique_lock<CCriticalSection> lock(m_critSection);
+    m_services.erase(std::remove_if(m_services.begin(), m_services.end(),
+                                    [&typeInfo](const auto& s)
+                                    {
+                                      const auto& sRef = *s.get();
+                                      return (typeid(sRef) == typeInfo);
+                                    }),
+                     m_services.end());
+  }
+
+private:
+  CCriticalSection m_critSection;
+  std::vector<std::shared_ptr<IPlatformService>> m_services; //!< Registered services
```

Comment:
This can be `std::unordered_map<std::type_index, std::shared_ptr<IPlatformService>>`
you get O(1) insert, add, delete and you know it is always only one per type, so you can't register 2 of the same type.

Reply:
Thanks for this very valuable feedback. Will change the code accordingly.